### ISSUE TITLE
Distributed: colons in numeric IPv6 address break format for specifying port number

### ISSUE THREAD

**Issue by mgkuhn** [JuliaLang/julia#34302]
`function launch_on_machine` in [managers.jl](https://github.com/JuliaLang/julia/blob/master/stdlib/Distributed/src/managers.jl) currently parses the address of a machine as
```julia
    machine_def = split(machine_bind[1], ':')
    # if this machine def has a port number, add the port information to the ssh flags
    if length(machine_def) > 2
        throw(ArgumentError("invalid machine definition format string: invalid port format \"$machine_def\""))
    end
    host = machine_def[1]
    portopt = ``
    if length(machine_def) == 2
        portstr = machine_def[2]
        if !all(isdigit, portstr) || (p = parse(Int,portstr); p < 1 || p > 65535)
            msg = "invalid machine definition format string: invalid port format \"$machine_def\""
            throw(ArgumentError(msg))
        end
        portopt = ` -p $(machine_def[2]) `
    end
```

So if the machine string contains a `:` character, the function assumes that this separates a host address from a port number. However, numeric IPv6 addresses also contain lots of colons!

The URL syntax had the same problem, and there [RFC 2732](https://tools.ietf.org/html/rfc2732) introduced the (somewhat cumbersome) convention/fix that in order to provide both a numeric IPv6 address and a port number separated by a colon, the former has to be enclosed in square brackets, as in `http://[2001:db8::1]:80/index.html`.

How do we want to fix this in `Distributed`:
1. Support the RFC 2732 square-bracket notation? This would be fully backwards compatible and is well known from URLs.
2. Switch to a different separator character (`Distributed` already uses `#` when reporting the bind address and port back to the caller)?

This affects the `machine_spec` parameter of [`addprocs`](https://docs.julialang.org/en/v1/stdlib/Distributed/#Distributed.addprocs), which can hold a colon-separated port number after both the `host` and `bind_addr` field.

**Comment by JeffBezanson** [comment]
Supporting RFC 2732 would be good :+1:

**Comment by mgkuhn** [comment]
RFC 2732 syntax is definitely the way to go: OpenSSH also uses it now in several command-line options. I have now implemented support for that syntax.

However, while testing it, I stumbled over several other parts of `Distributed` that are still very much hardwired to IPv4, for example `socket_reuse_port` and `bind_client_port` in [managers.jl](https://github.com/JuliaLang/julia/blob/master/stdlib/Distributed/src/managers.jl), functions last touched by @amitmurthy and @vtjnash. (The latter has already added a comment that in particular `bind_client_port` really belong into `Sockets`.) It clearly looks like `Distributed` has so far never been used with IPv6 connections.
